### PR TITLE
[direct call] Fix hang when caller id changes for actor task submission

### LINKED ISSUE #6329: Python script hangs when passing an actor as a parameter under direct call mode
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:  Linux
- **Ray installed from (source or binary)**: source
- **Ray version**: b7b655c85112abd20a8c92dac4b6062681ffbd19
- **Python version**: 3.7.2
- **Exact command to reproduce**:

### Describe the problem

The below script will hang with `export RAY_FORCE_DIRECT=1`.

### Source code / logs

```
import ray


@ray.remote
class Counter(object):
    def __init__(self, value):
        self.value = int(value)

    def increase(self, delta):
        self.value += int(delta)
        return self.value

@ray.remote
def foo(object):
  return ray.get(object.increase.remote(1))

@ray.remote
def bar(object):
  return ray.get(object.increase.remote(1))

@ray.remote
def zoo(object):
  return ray.get(object[0].increase.remote(1))

ray.init()
c = Counter.remote(0)
for _ in range(0, 100):
	print(ray.get(foo.remote(c)))
	print(ray.get(bar.remote(c)))
	print(ray.get(zoo.remote([c])))
```

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Actor tasks are assigned a sequence number per caller id. This is used to guarantee ordering on the receiver side. However, the caller id can change on the worker if the same worker gets passed the same actor handle multiple times.

This PR resolves the issue by decoupling the send position number from the sequence number. Now, it is an independent send_position counter.

## Related issue number

Closes https://github.com/ray-project/ray/issues/6329

### CODE REVIEW

**Comment by ericl** [comment]
Actually, it seems this counter is not used at all...

**Comment by stephanie-wang** [comment]
> Actually, it seems this counter is not used at all...

It was being used... tasks can get pushed into pending_tasks_ out-of-order depending on when their arguments get resolved. I think it would be fine if there was no timeout on the receiving end for out-of-order tasks, but as is, I don't think you can just remove the counter.

I think we have a test for out-of-order dependency resolution in direct_actor_transport_test, but maybe we should also add one in python?

**Comment by ericl** [comment]
Indeed that test is failing. Ok, the fix may be more complicated then.

**Comment by ericl** [comment]
Ok, should be fixed now.

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 1]
```diff
@@ -99,27 +99,32 @@ class CoreWorkerDirectActorTaskSubmitter {
...
   /// are ordered by the task number in the request.
   absl::flat_hash_map<ActorID, std::map<int64_t, std::unique_ptr<rpc::PushTaskRequest>>>
-      pending_requests_;
+      pending_requests_ GUARDED_BY(mu_);
 
-  /// Map from actor id to the sequence number of the next task to send to that
-  /// actor.
-  std::unordered_map<ActorID, int64_t> next_sequence_number_;
+  /// Map from actor id to the send position of the next task to queue for send
+  /// for that actor. This is always at or ahead of next_send_position_.
```
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_actor_transport.h]:
```suggestion
  /// for that actor. This is always greater than or equal to next_send_position_.
```

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 2]
```diff
@@ -99,27 +99,32 @@ class CoreWorkerDirectActorTaskSubmitter {
...
+  absl::flat_hash_map<ActorID, int64_t> next_send_position_to_assign_ GUARDED_BY(mu_);
+
+  /// Map from actor id to the send position of the next task to send to that actor.
+  /// Note that this differs from the sequence number in that it increases
```
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_actor_transport.h]:
```suggestion
  /// Note that this differs from the PushTaskRequest's sequence number in that it increases
```

**Review by stephanie-wang — APPROVED** [review]
I guess I should've added it in the original PR, but can you also add a comment explaining why it's necessary to enqueue tasks in order on the sender?

**Comment by ericl** [comment]
Done

**Comment by ericl** [comment]
I see F1203 06:01:24.969717 2846815040 node_manager.cc:2640]  Check failed: ready_task_id_set_copy.empty() 

Consistently on the OSX core worker tests in Travis. Any idea if this is fixed in master or is something else?

**Pull request merged**